pynumaflow/batchmapper/async_server.py

```diff
@@ -0,0 +1,107 @@
+import os
+
+import aiorun
+import grpc
+
+from pynumaflow._constants import (
+    MAX_MESSAGE_SIZE,
+    MAX_THREADS,
+    _LOGGER,
+    BATCH_MAP_SOCK_PATH,
+    BATCH_MAP_SERVER_INFO_FILE_PATH,
+)
+from pynumaflow.batchmapper._dtypes import BatchMapCallable
+from pynumaflow.batchmapper.servicer.async_servicer import AsyncBatchMapServicer
+from pynumaflow.proto.batchmapper import batchmap_pb2_grpc
+from pynumaflow.shared.server import NumaflowServer, start_async_server
+
+
+class BatchMapAsyncServer(NumaflowServer):
+    """
+    Class for a new Batch Map Async Server instance.
+    """
+
+    def __init__(
+        self,
+        batch_mapper_instance: BatchMapCallable,
+        sock_path=BATCH_MAP_SOCK_PATH,
+        max_message_size=MAX_MESSAGE_SIZE,
+        max_threads=MAX_THREADS,
+        server_info_file=BATCH_MAP_SERVER_INFO_FILE_PATH,
+    ):
+        """
+        Create a new grpc Async Batch Map Server instance.
+        A new servicer instance is created and attached to the server.
+        The server instance is returned.
+        Args:
+            batch_mapper_instance: The batch map stream instance to be used for Batch Map UDF
+            sock_path: The UNIX socket path to be used for the server
+            max_message_size: The max message size in bytes the server can receive and send
+            max_threads: The max number of threads to be spawned;
+                            defaults to number of processors x4
+
+        Example invocation:
+         class Flatmap(BatchMapper):
+            async def handler(
+                self,
+                datums: AsyncIterable[Datum],
+            ) -> BatchResponses:
+                batch_responses = BatchResponses()
+                async for datum in datums:
+                    val = datum.value
+                    _ = datum.event_time
+                    _ = datum.watermark
+                    strs = val.decode("utf-8").split(",")
+                    batch_response = BatchResponse.new_batch_response(datum.id)
+                    if len(strs) == 0:
+                        batch_response.append(Message.to_drop())
+                    else:
+                        for s in strs:
+                            batch_response.append(Message(str.encode(s)))
+                    batch_responses.append(batch_response)
+
+                return batch_responses
+
+        if __name__ == "__main__":
+            grpc_server = BatchMapAsyncServer(Flatmap())
+            grpc_server.start()
+        """
+        self.batch_mapper_instance: BatchMapCallable = batch_mapper_instance
+        self.sock_path = f"unix://{sock_path}"
+        self.max_threads = min(max_threads, int(os.getenv("MAX_THREADS", "4")))
```

Comment:
Let's have a max cap on the threads. Too many is not good. Let's keep a cap at 16

Reply:
Defaulting to 4, putting a max cap at 16
Will change for all server types